src/bio/bio_config.c

```diff
@@ -0,0 +1,531 @@
+		}
+
+		/* Move on to next subsystem config*/
+		ctx->config_it = spdk_json_next(ctx->config_it);
+	}
+out:
+	D_FREE(ctx->json_data);
+	D_FREE(ctx->values);
+	D_FREE(ctx);
+	return rc;
+}
+
+static int
+check_vmd_status(struct spdk_json_val *vmd_ss, bool *vmd_enabled)
+{
+	struct json_config_ctx	*ctx = calloc(1, sizeof(*ctx));
+	int			 rc = 0;
+
+	if (vmd_ss == NULL) {
+		return 0;
+	}
+	D_ASSERT(vmd_enabled != NULL);
+
+	if (!ctx) {
+		rc = -DER_NOMEM;
+		D_ERROR("Failed to allocate context, "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	/* Capture subsystem name and config array */
+	if (spdk_json_decode_object(vmd_ss, subsystem_decoders, SPDK_COUNTOF(subsystem_decoders),
+				    ctx)) {
+		D_ERROR("Failed to parse subsystem configuration\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	D_DEBUG(DB_MGMT, "subsystem '%.*s': found in JSON config\n", ctx->subsystem_name->len,
+		(char *)ctx->subsystem_name->start);
+
+	/* Get 'config' array first configuration entry */
+	ctx->config_it = spdk_json_array_first(ctx->config);
+
+	while (ctx->config_it != NULL) {
+		rc = load_vmd_subsystem_config(ctx, vmd_enabled);
+		if (rc != 0) {
+			goto out;
+		}
+
+		/* Move on to next subsystem config*/
+		ctx->config_it = spdk_json_next(ctx->config_it);
+	}
+out:
+	D_FREE(ctx->json_data);
+	D_FREE(ctx->values);
+	D_FREE(ctx);
+	return rc;
+}
+
+int
+bio_add_allowed_devices(const char *json_config_file, struct spdk_env_opts *opts)
+{
+	struct json_config_ctx	*ctx = calloc(1, sizeof(*ctx));
+	struct spdk_json_val	*bdev_ss = NULL, *vmd_ss = NULL;
+	bool			 vmd_enabled = false;
+	int			 rc = 0;
+
+	if (!ctx) {
+		rc = -DER_NOMEM;
+		D_ERROR("Failed to allocate context, "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	D_ASSERT(json_config_file);
+	rc = read_config(json_config_file, ctx);
+	if (rc) {
+		D_ERROR("config read failed");
+		return rc;
+	}
+
+	/* Capture subsystems array */
+	rc = spdk_json_find_array(ctx->values, "subsystems", NULL, &ctx->subsystems);
+	if (rc) {
+		D_ERROR("No 'subsystems' key JSON configuration file.\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	/* Get first subsystem */
+	ctx->subsystems_it = spdk_json_array_first(ctx->subsystems);
+	if (ctx->subsystems_it == NULL) {
+		D_ERROR("'subsystems' configuration is empty\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	while (ctx->subsystems_it != NULL) {
+		/* Capture subsystem name and config array */
+		if (spdk_json_decode_object(ctx->subsystems_it, subsystem_decoders,
+					    SPDK_COUNTOF(subsystem_decoders), ctx)) {
+			D_ERROR("Failed to parse subsystem configuration\n");
+			rc = -DER_INVAL;
+			goto out;
+		}
+
+		if (spdk_json_strequal(ctx->subsystem_name, "bdev"))
+			bdev_ss = ctx->subsystems_it;
+
+		if (spdk_json_strequal(ctx->subsystem_name, "vmd"))
+			vmd_ss = ctx->subsystems_it;
+
+		/* Move on to next subsystem */
+		ctx->subsystems_it = spdk_json_next(ctx->subsystems_it);
+	};
+
+	if (bdev_ss == NULL) {
+		D_WARN("JSON config missing bdev subsystem\n");
+		goto out;
```

Comment:
I wasn't sure, I decided to just put a warning and return zero because there might be cases where a missing bdev subsystem shouldn't be fatal. I think you are right and it is an error condition. A bdev subsystem with no attach methods is probably still valid (as a related aside) in the case of AIO I think.

Reply:
A spdk_nvme.conf without 'bdev' subsystem is at least invalid to DAOS (BIO), BIO will fail on not finding any devices.